core/client/src/client.rs

```diff
@@ -1477,21 +1355,202 @@ where
 	}
 }
 
-impl<B, E, Block, RA> ChainHead<Block> for Client<B, E, Block, RA>
+/// Implement Longest Chain Select implementation
+/// where 'longest' is defined as the highest number of blocks
+pub struct LongestChain<B, Block> {
+	backend: Arc<B>,
+	import_lock: Arc<Mutex<()>>,
```

Comment:
Given that `()` is obviously `Send` and `Sync`, there should be a comment here explaining what exactly the `Mutex` is protecting.  Even better would be to replace `()` by the actual type being protected.

Reply:
This is an ugly hack from back in the day when we split importing out of the client and all of the sudden we could do things in parallel with importing that we didn't want to do while importing. Before, the client would hold the "currently importing hash" but that wasn't possible anymore after, so we created a lock that is acquired when importing takes place -- and this particular implementation of longest chain needs to acquire that lock to work (unfortunately). As my focus was on the chain selection, I didn't attempt to fix this ugliness, but IMHO the lock should actually be in the `ImportQueue` and something you coordinate with it, rather than sharing the lock here.

Reply:
I'd put this into a ticket once we got this merged. Might be a good starter task.